### PR TITLE
Namespace(Schema) support + Catalog refactoring 

### PR DESCRIPTION
**Pull request by mengranwo**
####  1. catalog refactoring  
* Get rid of catalog table singleton. Make pg_table, pg_index, pg_attribute, pg_stats_tables, pg_trigger under per database. All of them are now stored in a per-database object in catalog_map (under catalog.h)
* Add version_id column into pg_table
* Add update(with index scan) helper function inside AbstractcCatalog.cpp  

####  2. Namespace(Schema) support
* [support create/drop schema](https://www.postgresql.org/docs/9.6/static/sql-createschema.html)

```
default_database=#CREATE TABLE foo(a int);
create table under default namespace "public"

default_database=#CREATE SCHEMA private;
default_database=#CREATE TABLE private.foo(a int);
create table with same name "foo" under different namespace "private"
default_database=#DROP SCHEMA private;
delete record from pg_namespace catalog table and drop table private.foo
```
* [Add pg_namespace catalog table](https://www.postgresql.org/docs/9.6/static/catalog-pg-namespace.html)
* Add namespace(schema) information in parser, binder, planner and executor  
* Need to pass in database name, schema name and table name to locate one unique table object

```
DataTable *table = catalog::Catalog::GetInstance()->GetTableWithName(
    database_name, schema_name, table_name, txn);

TableCatalogObject *table_object = catalog::Catalog::GetInstance()->GetTableObject(
    database_name, schema_name, table_name, txn);
```

####  3. To-Do
* <strike> Fix test cases (consult with PR #1286)
* Add more namespace test cases

### CODE REVIEW

File: test/statistics/stats_test.cpp [hunk 1]
```diff
@@ -157,10 +160,10 @@ TEST_F(StatsTests, MultiThreadStatsTest) {
   ASSERT_EQ(aggregated_stats.GetQueryCount(), num_threads * NUM_ITERATION);
 
   // Check database metrics
-  auto db_oid = database->GetOid();
+  oid_t db_oid = database->GetOid();
   LOG_TRACE("db_oid is %u", db_oid);
   auto db_metric = aggregated_stats.GetDatabaseMetric(db_oid);
-  ASSERT_EQ(db_metric->GetTxnCommitted().GetCounter(),
+  ASSERT_GT(db_metric->GetTxnCommitted().GetCounter(),
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply test/statistics/stats_test.cpp]: What stats test basically does is to start 8 threads, and within each thread invokes method **IncrementTxnCommitted()** 50 times, thus we will eventually have 8 * 50 = 400 committed transaction in that database. But we've changed table/index/query metrics catalog tables from global singleton to per database catalog table, every time we insert a stats record into corresponding catalog table, the number of committed transaction will be plus one. Since we don't have full control over how long the test will run and how many times the stats aggregator thread will be woken up, we can only have a nondeterministic number of committed transaction(but always larger than 400).

File: test/sql/timestamp_functions_sql_test.cpp [hunk 2]
```diff
@@ -80,19 +80,17 @@ TEST_F(TimestampFunctionsSQLTest, DatePartTest) {
   auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
   auto txn = txn_manager.BeginTransaction();
   catalog::Catalog::GetInstance()->CreateDatabase(DEFAULT_DB_NAME, txn);
-  catalog::Catalog::GetInstance()->Bootstrap();
+  // NOTE: Catalog::GetInstance()->Bootstrap() has been called in previous tests
+  // you can only call it once!
+  // catalog::Catalog::GetInstance()->Bootstrap();
   txn_manager.CommitTransaction(txn);
-  // Create a t
-  txn = txn_manager.BeginTransaction();
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply test/sql/timestamp_functions_sql_test.cpp]: cause within function **ExecuteSQLQuery** we'll handle the logic of create table and insert tuple(plus, the **txn** object has never been used)

File: test/optimizer/old_optimizer_test.cpp [hunk 3]
```diff
@@ -101,11 +101,6 @@ TEST_F(OldOptimizerTests, UpdateDelWithIndexScanTest) {
   traffic_cop.CommitQueryHelper();
 
   txn = txn_manager.BeginTransaction();
-  EXPECT_EQ(catalog::Catalog::GetInstance()
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply test/optimizer/old_optimizer_test.cpp]: Talk to Andy about this issue, he said it's fine to get rid of the **counting table/tableonject** tests. And the only exception is when deleting a table(make sure the table number is going down by 1)

File: test/function/functions_test.cpp [hunk 4]
```diff
@@ -31,16 +31,19 @@ class FunctionsTests : public PelotonTest {
...
+    // NOTE: Catalog::GetInstance()->Bootstrap() has been called in previous
+    // unit tests you can only call it once!
+    // auto catalog =
+    // catalog::Catalog::GetInstance(); catalog->Bootstrap();
   }
 };
 
 TEST_F(FunctionsTests, CatalogTest) {
   auto catalog = catalog::Catalog::GetInstance();
+  catalog->Bootstrap();
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply test/function/functions_test.cpp]: Fixed it!

File: test/statistics/stats_test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (poojanilangekar) [reply test/statistics/stats_test.cpp]: Actually, could you add this explanation as a comment? So if someone changes the test, they know what they should expect.

File: test/executor/create_index_test.cpp [hunk 5]
```diff
@@ -115,11 +115,6 @@ TEST_F(CreateIndexTests, CreatingIndex) {
   traffic_cop.CommitQueryHelper();
 
   txn = txn_manager.BeginTransaction();
-  EXPECT_EQ(catalog::Catalog::GetInstance()
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply test/executor/create_index_test.cpp]: Talk to Andy about this issue, he said it's fine to get rid of the counting table/tableonject tests. And the only exception is when deleting a table(make sure the table number is going down by 1)

File: src/storage/data_table.cpp [hunk 6]
```diff
@@ -13,11 +13,9 @@
 #include <mutex>
 #include <utility>
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/storage/data_table.cpp]: Yes we do change how we access the trigger catalog table.(not a singleton anymore)
line 1408

File: src/statistics/stats_aggregator.cpp [hunk 7]
```diff
@@ -329,16 +347,5 @@ void StatsAggregator::UnregisterContext(std::thread::id id) {
   }
 }
 
-storage::DataTable *StatsAggregator::GetMetricTable(std::string table_name) {
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/statistics/stats_aggregator.cpp]: This function has never been invoked.

File: src/include/storage/database.h [hunk 8]
```diff
@@ -54,9 +54,6 @@ class Database : public Printable {
   // Throw CatalogException if such table is not found
   storage::DataTable *GetTableWithOid(const oid_t table_oid) const;
 
-  // Throw CatalogException if such table is not found
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/storage/database.h]: We don't have a physical level object(like database.h or data_table.h) corresponds to **schema(namespace)**, we only store this information within pg_namespace and pg_table catalog table.

File: src/include/statistics/stats_aggregator.h [hunk 9]
```diff
@@ -89,9 +89,6 @@ class StatsAggregator {
   // instead of explicitly unregistering it.
   void UnregisterContext(std::thread::id id);
 
-  // Utility function to get the metric table
-  storage::DataTable *GetMetricTable(std::string table_name);
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/statistics/stats_aggregator.h]: This function has never been invoked. It used to be a private helper function within stats aggregator, but no one is calling this function anymore(even before this PR).

File: src/binder/bind_node_visitor.cpp [hunk 10]
```diff
@@ -155,8 +155,8 @@ void BindNodeVisitor::Visit(parser::UpdateStatement *node) {
 void BindNodeVisitor::Visit(parser::DeleteStatement *node) {
   context_ = std::make_shared<BinderContext>(nullptr);
   node->TryBindDatabaseName(default_database_name_);
-  context_->AddRegularTable(node->GetDatabaseName(), node->GetTableName(),
-                            node->GetTableName(), txn_);
+  context_->AddRegularTable(node->GetDatabaseName(), node->GetSchemaName(),
+                            node->GetTableName(), node->GetTableName(), txn_);
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/binder/bind_node_visitor.cpp]: The second table name is **const std::string table_alias**, I didn't change the logic here except adding namespace parameter(I'll confirm with bowei)

File: src/binder/bind_node_visitor.cpp [hunk 11]
```diff
@@ -174,8 +174,8 @@ void BindNodeVisitor::Visit(parser::CreateStatement *node) {
 void BindNodeVisitor::Visit(parser::InsertStatement *node) {
   node->TryBindDatabaseName(default_database_name_);
   context_ = std::make_shared<BinderContext>(nullptr);
-  context_->AddRegularTable(node->GetDatabaseName(), node->GetTableName(),
-                            node->GetTableName(), txn_);
+  context_->AddRegularTable(node->GetDatabaseName(), node->GetSchemaName(),
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/binder/bind_node_visitor.cpp]: See above

File: src/include/catalog/index_metrics_catalog.h [hunk 12]
```diff
@@ -39,37 +38,33 @@ namespace catalog {
...
-
   enum ColumnId {
-    DATABASE_OID = 0,
-    TABLE_OID = 1,
-    INDEX_OID = 2,
-    READS = 3,
-    DELETES = 4,
-    INSERTS = 5,
-    TIME_STAMP = 6,
+    TABLE_OID = 0,
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/catalog/index_metrics_catalog.h]: Table oid is going to be enough(we don't have same table oid under different namespace, even they may have the same table name).

File: src/include/catalog/system_catalogs.h [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+}  // namespace storage
+
+namespace catalog {
+class DatabaseCatalog;
+class SchemaCatalog;
+class TableCatalog;
+class IndexCatalog;
+class ColumnCatalog;
+
+class SystemCatalogs {
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/catalog/system_catalogs.h]: Added!

File: src/include/catalog/system_catalogs.h [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  QueryMetricsCatalog *GetQueryMetricsCatalog() {
+    if (!pg_query_metrics) {
+      throw CatalogException("Query metrics catalog has not been initialized");
+    }
+    return pg_query_metrics;
+  }
+
+ private:
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/catalog/system_catalogs.h]: Done!

File: src/include/catalog/schema_catalog.h [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+  friend class DatabaseCatalogObject;
+
+ public:
+  SchemaCatalogObject(executor::LogicalTile *tile,
+                      concurrency::TransactionContext *txn);
+
+  inline oid_t GetSchemaOid() { return schema_oid; }
+  inline const std::string &GetSchemaName() { return schema_name; }
+
+ private:
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/catalog/schema_catalog.h]: Done!

File: src/include/catalog/column_catalog.h [hunk 16]
```diff
@@ -83,10 +81,11 @@ class ColumnCatalog : public AbstractCatalog {
   // write Related API
   //===--------------------------------------------------------------------===//
   bool InsertColumn(oid_t table_oid, const std::string &column_name,
-                    oid_t column_id, oid_t column_offset,
+                    uint32_t column_id, uint32_t column_offset,
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/catalog/column_catalog.h]: Done!

File: src/include/planner/plan_util.h [hunk 17]
```diff
@@ -54,11 +54,10 @@ class PlanUtil {
...
   /**
-  * @brief Get the indexes affected by a given query
-  * @param CatalogCache
-  * @param SQLStatement
-  * @return set of affected index object ids
-  */
+   * @brief Get the indexes affected by a given query
+   * @param SQLStatement
+   * @return set of affected index object ids
+   */
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/planner/plan_util.h]: Reverted! Thanks for pointing that out, I didn't understand why we need **CatalogCache** in the beginning, so I just deleted it and forgot to add it back.

File: src/include/catalog/query_metrics_catalog.h [hunk 18]
```diff
@@ -63,19 +59,18 @@ class QueryMetricsCatalog : public AbstractCatalog {
...
-      const std::string &name, oid_t database_oid,
-      concurrency::TransactionContext *txn);
-  int64_t GetNumParams(const std::string &name, oid_t database_oid,
+      const std::string &name, concurrency::TransactionContext *txn);
+  int64_t GetNumParams(const std::string &name,
                        concurrency::TransactionContext *txn);
-  // TODO: add more if needed
-
+  // TODO: In theory, we don't need database_oid
+  // but now we store all the query metrics under default database "pelton"
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/catalog/query_metrics_catalog.h]: Done!

File: src/include/catalog/table_metrics_catalog.h [hunk 19]
```diff
@@ -39,37 +38,33 @@ namespace catalog {
...
   //===--------------------------------------------------------------------===//
   // Read-only Related API
   //===--------------------------------------------------------------------===//
   // TODO: add if needed
 
  private:
-  TableMetricsCatalog(concurrency::TransactionContext *txn);
-
   enum ColumnId {
-    DATABASE_OID = 0,
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/include/catalog/table_metrics_catalog.h]: Table oid is going to be enough(we don't have same table oid under different namespace, even they may have the same table name).

File: src/catalog/abstract_catalog.cpp [hunk 20]
```diff
@@ -45,84 +46,95 @@ AbstractCatalog::AbstractCatalog(oid_t catalog_table_oid,
...
+    columns.push_back(tuple->GetSchema()->GetColumn(i).GetName());
+    values[0].emplace_back(
+        new expression::ConstantValueExpression(tuple->GetValue(i)));
+  }
+  auto node =
+      std::make_shared<planner::InsertPlan>(catalog_table_, &columns, &values);
 
-  return status;
+  executor::ExecutionResult this_p_status;
+  auto on_complete = [&this_p_status](
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/catalog/abstract_catalog.cpp]: We are trying to convert the old execution engine to llvm execution engine. 
Since the llvm support "insert" functionality, so we decided to change the logic. (It may not be useful in the future, cause I know there is a team working on generating pre-compiled plan for catalog)

File: src/catalog/abstract_catalog.cpp [hunk 21]
```diff
@@ -286,5 +298,70 @@ void AbstractCatalog::AddIndex(const std::vector<oid_t> &key_attrs,
...
+  if (txn == nullptr) throw CatalogException("Scan table requires transaction");
+
+  std::unique_ptr<executor::ExecutorContext> context(
+      new executor::ExecutorContext(txn));
+  // Construct index scan executor
+  auto index = catalog_table_->GetIndex(index_offset);
+  std::vector<oid_t> key_column_offsets =
+      index->GetMetadata()->GetKeySchema()->GetIndexedColumns();
+  PELOTON_ASSERT(scan_values.size() == key_column_offsets.size());
+  std::vector<ExpressionType> expr_types(scan_values.size(),
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/catalog/abstract_catalog.cpp]: Done!

File: src/catalog/catalog.cpp [hunk 22]
```diff
@@ -46,114 +47,152 @@ Catalog *Catalog::GetInstance() {
...
+  // Commit transaction
+  txn_manager.CommitTransaction(txn);
+}
+
+/* This function *MUST* be called after a new database is created to bootstrap
+ * all system catalog tables for that database.
+ * The system catalog tables must be created in certain order to make sure
+ * all tuples are indexed (actually this might be fine now after Paulo's fix)
+ */
+void Catalog::BootstrapSystemCatalogs(storage::Database *database,
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/catalog/catalog.cpp]: done!

File: src/catalog/table_catalog.cpp [hunk 23]
```diff
@@ -541,5 +561,29 @@ TableCatalog::GetTableObjects(oid_t database_oid,
   return database_object->GetTableObjects();
 }
 
+bool TableCatalog::UpdateVersionId(oid_t update_val, oid_t table_oid,
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/catalog/table_catalog.cpp]: done!

File: src/parser/postgresparser.cpp [hunk 24]
```diff
@@ -1418,8 +1448,18 @@ parser::DropStatement *PostgresParser::DropIndexTransform(DropStmt *root) {
   auto result = new DropStatement(DropStatement::EntityType::kIndex);
   auto cell = root->objects->head;
   auto list = reinterpret_cast<List *>(cell->data.ptr_value);
-  result->SetIndexName(
-      reinterpret_cast<value *>(list->head->data.ptr_value)->val.str);
+  // if schema name is specified
+  if (list->length == 2) {
```
[the remark this answers is not in the document]
Comment (mengranwo) [reply src/parser/postgresparser.cpp]: Done!

**Comment by mengranwo** [comment]
Hey @poojanilangekar!  
Thank you for your reviews and I've made changes according to your suggestion. The following is a summary to what I've modified and answers to some of your question.  

* delete all the commented out line
* Add java doc style documentation if necessary
* Add CATALOG_TABLES_COUNT variable into catalog_defaults.h to get rid of the magic number 8
*  The reason why we change the visibility of the catalog contractors from private to public is that: before this PR each catalog table is a global singleton (thus the constructor must be private), but now each database has their own set of catalog tables (a.k.a system_catalogs.h), when creating a new system catalog object, we need to create multiple catalog tables(pg_table, pg_index, etc), thus the constructors must be public now  
* In all the test cases, we require you to use lowercase database_name. Cause many test cases invoke catalog::CreateDatabase(database_name) directly, and we use create sql to build secondary catalog tables(e.g CREATE TABLE database_name.pg_trigger(a int);) which means that this sql statement will pass through parser--> binder-->optimizer-->planner-->executor.And binder/optimizer will convert whatever name to be lowercase. We need to keep both name in sync, otherwise "can't find database" exception will be thrown.
* PR # 1286 is adding TempSchema
* I'll push the schema test cases during weekend.

File: test/statistics/stats_test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mengranwo) [reply test/statistics/stats_test.cpp]: Sure!

File: test/optimizer/optimizer_rule_test.cpp [hunk 25]
```diff
@@ -23,6 +23,8 @@
 #include "executor/update_executor.h"
 #include "expression/abstract_expression.h"
 #include "expression/operator_expression.h"
+
+#define private public
```
[the remark this answers is not in the document]
Comment (poojanilangekar) [reply test/optimizer/optimizer_rule_test.cpp]: @mengranwo I think you missed this one. Can you please change this as well?

**Comment by poojanilangekar** [comment]
@tcm-marcel, The same issue again. Travis build is not triggered. Do you think the bug is because of some error in our configuration or a bug on their end?

File: test/catalog/catalog_test.cpp [hunk 26]
```diff
@@ -116,40 +117,6 @@ TEST_F(CatalogTests, CreatingTable) {
                         ->GetColumnObject(1)
                         ->GetColumnName());
   txn_manager.CommitTransaction(txn);
-  // EXPECT_EQ(5, time_stamp);
```
Comment (poojanilangekar) [inline test/catalog/catalog_test.cpp]: Would it be possible for you to check the table count? Currently it should be 8, correct? Similarly to the index as well? 
So this way we can check that people always account for the additional catalog tables/indexes they create.

File: test/sql/testing_sql_util.cpp [hunk 27]
```diff
@@ -248,7 +249,9 @@ void TestingSQLUtil::ExecuteSQLQueryAndCheckResult(
   // Execute query
   TestingSQLUtil::ExecuteSQLQuery(std::move(query), result, tuple_descriptor,
                                   rows_changed, error_message);
-  unsigned int rows = result.size() / tuple_descriptor.size();
+  unsigned int rows = tuple_descriptor.size() == 0
```
Comment (poojanilangekar) [inline test/sql/testing_sql_util.cpp]: Should we add a `PELOTON_ASSERT` instead? I mean, ideally we should only have an empty`result` if the `tuple_descriptor.size()` is 0, right? Or am I missing a corner case?

File: test/optimizer/optimizer_rule_test.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (poojanilangekar) [reply test/optimizer/optimizer_rule_test.cpp]: Please change this as well.

**Comment by camellyx** [comment]
@poojanilangekar Thanks for your comments. We are working on adding more tests, if possible, once we have time.

**Comment by poojanilangekar** [comment]
@camellyx Can this PR be merged in for now?
You can add more tests in the subsequent PR. If this gets merged in and I merge in the layout PR #1327, then we can start work on schema change. Or do you want to wait? Let me know what works for you.

**Comment by mengranwo** [comment]
@poojanilangekar I think we are ready to go.

**Comment by poojanilangekar** [comment]
Thanks @mengranwo!
@pervazea @apavlo Can you please merge this in?

**Comment by nwang57** [comment]
@mengranwo I think there are typos in src/include/catalog/catalog_defaults.h and other places. `DEFUALT_SCHEMA_OID` and `DEFUALT_SCHEMA_NAME ` which should be `DEFAULT_SCHEMA_OID` and `DEFAULT_SCHEMA_NAME`
